### PR TITLE
Closes #559: Dataset Embargo - added GUI support

### LINKED ISSUE #559: User with edit dataset permission should be able to assign an embargo for dataset

### PR DESCRIPTION
**Pull request by dbojanek**

### CODE REVIEW

File: dataverse-persistence/src/main/resources/Bundle_en.properties [hunk 1]
```diff
@@ -1439,6 +1440,24 @@ dataset.privateurl.roleassigeeTitle=Private URL Enabled
 dataset.privateurl.createdSuccess=Success!
 dataset.privateurl.disabledSuccess=You have successfully disabled the Private URL for this dataset.
 dataset.privateurl.noPermToCreate=To create a Private URL you must have the following permissions: {0}.
+dataset.embargo.title=Embargo
+dataset.embargo.description=Date for making published dataset files available.
+dataset.embargo.header=Dataset Embargo Date
+dataset.embargo.tip=Hide files in your dataset until embargo date.
+dataset.embargo.message=Currently embargo is set to : {0}
```
Comment (wtr-icm) [inline dataverse-persistence/src/main/resources/Bundle_en.properties:1447]: Why the space before ":"? (Also in polish bundle).

File: dataverse-persistence/src/main/resources/Bundle_pl.properties [hunk 2]
```diff
@@ -1440,6 +1441,24 @@ dataset.privateurl.roleassigeeTitle=Prywatny URL
...
+dataset.embargo.description=Data udost\u0119pnienia plik\u00F3w w opublikowanym zbiorze danych.
+dataset.embargo.header=Data embarga zbioru danych
+dataset.embargo.tip=Ukryj pliki w opublikowanym zbiorze danych do daty embarga.
+dataset.embargo.message=Zapisane embargo przypada na dat\u0119 : {0}
+dataset.embargo.filesTab.message.empty=Pliki b\u0119d\u0105 dost\u0119pne od dnia: {0}
+dataset.embargo.setEmbargo=Na\u0142\u00F3\u017C embargo
+dataset.embargo.lift.button=Zdejmij embargo
+dataset.embargo.lift.confirmButton=Tak, zdejmij embargo
+dataset.embargo.lift.confirmationText=Czy na pewno chcesz zdj\u0105\u0107 embargo? Pliki w Twoim zbiorze danych b\u0119d\u0105 widoczne dla wszystkich u\u017Cytkownik\u00F3w.
+dataset.embargo.lift.publishedConfirmationText=Once you lift the embargo, you will not be able to set it again.
```
Comment (wtr-icm) [inline dataverse-persistence/src/main/resources/Bundle_pl.properties:1454]: Translation missing.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 3]
```diff
@@ -1203,4 +1222,55 @@ public boolean isSameTermsOfUseForAllFiles() {
...
+    public void updateEmbargoDate() {
+        Try.of(() -> datasetService.setDatasetEmbargoDate(dataset, currentEmbargoDate))
+                .onSuccess(ds -> JsfHelper.addFlashSuccessMessage(BundleUtil.getStringFromBundle("dataset.embargo.save.successMessage")))
+                .onFailure(ds -> JsfHelper.addFlashErrorMessage(BundleUtil.getStringFromBundle("dataset.embargo.save.failureMessage")));
+    }
+
+    public void liftEmbargo() {
+        Try.of(() -> datasetService.liftDatasetEmbargoDate(dataset))
+                .onSuccess(ds -> currentEmbargoDate = null)
+                .onSuccess(ds -> JsfHelper.addFlashSuccessMessage(BundleUtil.getStringFromBundle("dataset.embargo.lift.successMessage")))
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:1246]: Why not in one call to ```onSuccess```?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 4]
```diff
@@ -1203,4 +1222,55 @@ public boolean isSameTermsOfUseForAllFiles() {
...
+                .onFailure(ds -> JsfHelper.addFlashErrorMessage(BundleUtil.getStringFromBundle("dataset.embargo.lift.failureMessage")));
+    }
+
+    public String getCurrentEmbargoDateForDisplay() {
+        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: String ```"yyyy-MM-dd"``` should be extracted into a constant. Probably in separate file, as it is also used in other places.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 5]
```diff
@@ -1203,4 +1222,55 @@ public boolean isSameTermsOfUseForAllFiles() {
...
+        return format.format(currentEmbargoDate);
+    }
+
+    public void validateEmbargoDate(FacesContext context, UIComponent toValidate, Object embargoDate) {
+        if(!Objects.isNull(embargoDate) &&
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: Why not ```embargoDate != null```?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/tab/DatasetFilesTab.java [hunk 6]
```diff
@@ -668,6 +668,11 @@ public boolean isAllFilesSelected() {
         return selectedFiles.size() >= allCurrentFilesCount();
     }
 
+    public String getEmbargoDateForDisplay() {
+        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/tab/DatasetFilesTab.java]: Date format string should be replaced by constant.

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java [hunk 7]
```diff
@@ -75,4 +92,61 @@ public void removeDatasetThumbnail() {
...
+    @Test
+    public void setDatasetEmbargoDate_lockedDataset() {
+        // given
+        Dataset dataset = new Dataset();
+        DatasetLock lock = new DatasetLock(DatasetLock.Reason.InReview, MocksFactory.makeAuthenticatedUser("Jurek", "Kiler"));
+        lock.setId(1L);
+        dataset.addLock(lock);
+        Date embargoDate = Date.from(Instant.now().truncatedTo(ChronoUnit.DAYS).plus(2, ChronoUnit.DAYS));
+
+        // when
```
Comment (wtr-icm) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java]: Suggestion: better "when & then" as we have an assertion coupled with service invocation.

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java [hunk 8]
```diff
@@ -75,4 +92,61 @@ public void removeDatasetThumbnail() {
...
+        Exception exception = Assertions.assertThrows(IllegalStateException.class, () -> {
+            datasetService.setDatasetEmbargoDate(dataset, embargoDate);
+        });
+
+        // then
+        String message = "Update embargo date failed. Dataset is locked. [DatasetLock[ id=1 ]]";
+        Assertions.assertEquals(message, exception.getMessage());
```
Comment (wtr-icm) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java]: In this case we should either compare only invariant part of the message: extract it to the constant in the place where it's defined and use this constant also here or create a method that produces message for exception and call it within test to produce expected message. 
Currently any change to message or ```DatasetLock#toString()``` will break the test.

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java [hunk 9]
```diff
@@ -75,4 +92,61 @@ public void removeDatasetThumbnail() {
...
+
+        Date embargoDate = Date.from(Instant.now().truncatedTo(ChronoUnit.DAYS).plus(2, ChronoUnit.DAYS));
+
+        // when
+        Exception exception = Assertions.assertThrows(IllegalStateException.class, () -> {
+            datasetService.setDatasetEmbargoDate(dataset, embargoDate);
+        });
+
+        // then
+        String message = "Setting embargo date failed. Dataset is in wrong state.";
```
Comment (wtr-icm) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetServiceTest.java]: As above.

**Review by wtr-icm — CHANGES_REQUESTED** [review]

File: dataverse-persistence/src/main/resources/Bundle_en.properties (hunk 1 above)
Comment (dbojanek) [reply dataverse-persistence/src/main/resources/Bundle_en.properties:1447]: Ask @wfenrich :)
This is copy-pasted from what he'd made.

File: dataverse-persistence/src/main/resources/Bundle_pl.properties (hunk 2 above)
Comment (dbojanek) [reply dataverse-persistence/src/main/resources/Bundle_pl.properties:1454]: As discussed on daily - text were fixed in separate branch.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 3 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:1246]: While not official standard, we tend to use this form whenever:
1) there are only few success/failure operations
2) you can divide them into blocks (here one success is for object modification and the other for banner message)

But yet again, it's more of a gentlemen's agreement than a standard.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 4 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: I'll ask @madryk if we want to put it in app properties.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 3 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:1246]: Ok.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 4 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: Adding app-wide property "DefaultDateFormat".

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/tab/DatasetFilesTab.java (hunk 6 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/tab/DatasetFilesTab.java]: `SimpleDateFormat format = new SimpleDateFormat(settingsService.getValueForKey(SettingsServiceBean.Key.DefaultDateFormat));` now

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 5 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: This was copied from some other validate method. I didn't think about it but sure, i can simplify this to `!=`

**Review by wtr-icm — APPROVED** [review]

**Pull request merged**